### PR TITLE
Migrate deployment processors to event sourcing

### LINKED ISSUE #6173: Apply event-sourcing to the deployment processors
**Description**

* apply the event-sourcing concept to the deployment processors

Open question:
* How do we keep track if a deployment is distributed or needs to be distributed again? (pending deployments in state)

Reference:
* [ZEP: Deployment](https://github.com/zeebe-io/enhancements/blob/master/ZEP004-wf-stream-processing.md#deploy-workflow)

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

**NOTE:** This PR is based on https://github.com/zeebe-io/zeebe/pull/6456 so we should merge this first. Or we just merge  this and close the other one :shrug: 

Last part of migrating the deployment processors to the event sourcing.

This PR does:

 * marking Deployment processors as migrated
 * remove unused code
 * remove mutable state usage from processor
      * version is requested on transforming and incremented 
      * version is set on appliers (on put workflow)
 * add test for deployment on multiple partitions
 * multiple deployment distribution complete commands are rejected now (caused before multiple FULLY_DISTRIBUTED events)
 * 

## Related issues

closes #6173 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java [hunk 1]
```diff
@@ -132,21 +124,8 @@ private void persistWorkflow(final long workflowKey, final WorkflowRecord workfl
...
-      final DeploymentResource resource) {
-    persistedWorkflow.wrap(resource, workflowRecord, workflowKey);
-    this.workflowKey.wrapLong(workflowKey);
-    workflowColumnFamily.put(this.workflowKey, persistedWorkflow);
 
-    workflowId.wrapBuffer(workflowRecord.getBpmnProcessIdBuffer());
-    workflowVersion.wrapLong(workflowRecord.getVersion());
-
-    workflowByIdAndVersionColumnFamily.put(idAndVersionKey, persistedWorkflow);
+    versionManager.setValue(workflowRecord.getBpmnProcessId(), workflowRecord.getVersion());
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java:128]: It seems that we have two column families that stores the latest version of the workflow: `WORKFLOW_VERSION` (`versionManager`) and `WORKFLOW_CACHE_LATEST_KEY` (`latestWorkflowColumnFamily`) :thinking: 
Can we get rid of one of these column families?

File: engine/src/test/java/io/zeebe/engine/processing/deployment/MultiPartitionDeploymentLifecycleTest.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            RecordType.COMMAND,
+            RecordType.EVENT,
+            RecordType.EVENT,
+            RecordType.EVENT,
+            RecordType.EVENT,
+            RecordType.COMMAND,
+            RecordType.COMMAND,
+            RecordType.EVENT,
+            RecordType.EVENT,
+            RecordType.EVENT);
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processing/deployment/MultiPartitionDeploymentLifecycleTest.java:82]: We can combine the two verifications using a `tuple` of `<RecordType, Intent>`.

**Review by saig0 — COMMENTED** [review]
Looks good so far :fish_cake: 

Please have a look at my comments. I'll do the final review after #6456 is merged.

File: engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java (hunk 1 above)
Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java:128]: ok this is weird. why we have that?

Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java:128]: can we do this separate I feel uncomfortable with that :see_no_evil:

**Comment by Zelldon** [comment]
I rebased the branch with develop after the other PR was merged. Please have another look @saig0 I adjusted already the test.

File: engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java (hunk 1 above)
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/state/deployment/DbWorkflowState.java:128]: :laughing: Sure. We can remove the column in a follow-up PR :heavy_check_mark:

**Review by saig0 — APPROVED** [review]
:tada:

**Review by saig0 — APPROVED** [review]

**Comment by Zelldon** [comment]
Awesome thanks for the fast review @saig0 :cookie: 

bors r+

**Pull request merged**